File: packages/python/src/xenoteer/examples/phase6_behaviors.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0
"""Qualify ten public SDK behaviors against the deterministic desktop fixture.

Run this installed-package example with the environment documented in the
package README. The Phase 6 artifact gate supplies those values and launches
the GTK fixture from the exact derived image before invoking this module.
"""

from __future__ import annotations

import asyncio
import hashlib
import os
import re
import sys
from collections.abc import Mapping
from pathlib import Path
from typing import Any, cast

import xenoteer
from xenoteer import ArtifactRef, ClientOptions, XenoteerClient, XenoteerError


BEHAVIORS = (
    "status-capabilities",
    "scoped-lease-fixture-launch",
    "exact-window-element",
    "semantic-invoke",
    "smooth-physical-click-postcondition",
    "unicode-text-strategy",
    "screenshot-on-failure",
    "reconnect-known-command",
    "stale-reference-restart",
    "view-only-browser-ticket",
)
GTK_TITLE = "Xenoteer GTK3 Fixture — Main"
XMESSAGE_TITLE = "xmessage"
VIEWER_ORIGIN = "https://viewer.example"
UNICODE_TEXT = "Xenoteer — العربية — 中文 — e\u0301 — 😀"


def required(name: str) -> str:
    value = os.environ.get(name)
    if not value:
        raise RuntimeError(f"required environment is missing: {name}")
    return value


def require(condition: bool, message: str) -> None:
    if not condition:
        raise RuntimeError(message)


def verify_installed_origin() -> None:
    expected = Path(required("XENOTEER_EXPECTED_INSTALL_ROOT")).resolve(strict=True)
    module_file = Path(xenoteer.__file__).resolve(strict=True)
    require(
        module_file.is_relative_to(expected),
        "Python SDK resolved outside the staged distribution installation",
    )


def marker(language: str, behavior: str) -> None:
    print(f"quickstart-ok language={language} behavior={behavior}")


def window_selector(title: str) -> dict[str, Any]:
    return {
        "type": "predicate",
        "predicate": {
            "type": "text",
            "field": "title",
            "matcher": {
                "type": "exact",
                "value": title,
                "case_sensitive": True,
            },
        },
    }


def element_selector(name: str) -> dict[str, Any]:
    return {
        "scope": {"type": "desktop"},
        "predicates": [
            {
                "type": "name",
                "matcher": {
                    "type": "exact",
                    "value": name,
                    "case_sensitive": True,
                },
            }
        ],
        "order": "object_path_ascending",
        "result_index": None,
    }


def accessibility_options(*, component: bool = False) -> dict[str, Any]:
    return {
        "expansion": {
            "actions": False,
            "value": False,
            "text_metadata": False,
            "text_content": False,
            "attributes": False,
            "relations": False,
            "component": component,
        },
        "limits": {
            "max_visited_nodes": 25_000,
            "max_depth": 64,
            "max_matches": 10,
            "timeout_ms": 10_000,
        },
    }


async def wait_element(desktop: Any, name: str) -> Any:
    selector = element_selector(name)
    await desktop.accessibility.wait(
        {
            "target": {
                "type": "selector",
                "selector": selector,
                "quantifier": "exactly_one",
            },
            "predicate": {"type": "exists"},
            "after_revision": None,
            "timeout_ms": 30_000,
            "allow_poll_fallback": True,
            **accessibility_options(component=False),
        }
    )
    return await desktop.accessibility.one(
        selector,
        **accessibility_options(component=True),
    )


async def wait_window(desktop: Any, title: str) -> Any:
    selector = window_selector(title)
    await desktop.windows.wait(
        {
            "target": {
                "type": "selector",
                "selector": selector,
                "quantifier": "exactly_one",
            },
            "predicate": {"type": "exists"},
            "after_revision": None,
            "timeout_ms": 30_000,
        }
    )
    return await desktop.windows.one(selector, order="creation_ascending")


async def terminal(submission: Any, label: str) -> dict[str, Any]:
    handle = await submission
    result = await handle.wait_until_terminal(20)
    if not isinstance(result, dict) or any(not isinstance(key, str) for key in result):
        raise RuntimeError(f"{label} returned an invalid terminal result")
    typed_result = cast(dict[str, Any], result)
    require(
        typed_result.get("lifecycle") == "succeeded",
        f"{label} did not succeed",
    )
    return typed_result


def outcome(result: Mapping[str, Any], kind: str) -> Mapping[str, Any]:
    value = result.get("outcome")
    if not isinstance(value, Mapping) or value.get("type") != kind:
        raise RuntimeError(f"command omitted its {kind} outcome")
    return value


async def launch_xmessage(desktop: Any, message: str) -> Mapping[str, Any]:
    result = await terminal(
        desktop.applications.launch(
            "xmessage",
            [message],
        ),
        "xmessage launch",
    )
    process = outcome(result, "application_launched").get("process")
    if not isinstance(process, Mapping):
        raise RuntimeError("launch outcome omitted its exact process")
    return process


async def terminate_process(desktop: Any, process: Mapping[str, Any]) -> None:
    result = await terminal(
        desktop.applications.terminate(process, grace=2),
        "fixture termination",
    )
    terminated = outcome(result, "process_terminated").get("process")
    require(
        isinstance(terminated, Mapping) and terminated.get("state") == "exited",
        "fixture termination did not reap the exact process",
    )


async def exercise_connected(client: XenoteerClient, language: str) -> None:
    desktop = client.desktop()
    capabilities = client.status.capabilities.get("capabilities")
    if not isinstance(capabilities, list):
        raise RuntimeError("status omitted capabilities")
    available = {
        item.get("id")
        for item in capabilities
        if isinstance(item, Mapping) and item.get("status") == "available"
    }
    required_capabilities = {
        "accessibility.atspi",
        "capture.screenshot",
        "input.pointer.smooth",
        "process.managed.terminate",
        "viewer.novnc.view_only",
        "window.observe.wait",
    }
    require(
        required_capabilities <= available,
        "fixture did not advertise every behavior capability as available",
    )
    marker(language, BEHAVIORS[0])

    async with desktop.control(ttl=60) as lease:
        await exercise_scoped(desktop, language, lease)


async def exercise_scoped(desktop: Any, language: str, lease: Any) -> None:
    message = f"Xenoteer SDK Phase 6 — {language}"
    process: Mapping[str, Any] | None = None
    screenshot_artifact: ArtifactRef | None = None
    operation_failure: BaseException | None = None
    try:
        process = await launch_xmessage(desktop, message)
        marker(language, BEHAVIORS[1])

        xmessage_window = await wait_window(desktop, XMESSAGE_TITLE)
        gtk_window = await wait_window(desktop, GTK_TITLE)
        button = await wait_element(desktop, "Stable Button")
        entry = await wait_element(desktop, "Stable Entry")
        require(xmessage_window.identity != gtk_window.identity, "exact windows aliased")
        require(button.identity != entry.identity, "exact elements aliased")
        correlated = await xmessage_window.snapshot()
        require(
            correlated.get("window", {})
            .get("snapshot", {})
            .get("process", {})
            .get("managed_process")
            == process,
            "exact xmessage window did not correlate to the launched process",
        )
        marker(language, BEHAVIORS[2])

        invoke_result = await terminal(button.invoke(), "semantic invoke")
        invoke_outcome = outcome(invoke_result, "element_action")
        invoke_evidence = invoke_outcome.get("result")
        require(
            isinstance(invoke_evidence, Mapping)
            and invoke_evidence.get("operation") == "invoke"
            and invoke_evidence.get("element") == button.identity
            and isinstance(invoke_evidence.get("evidence"), Mapping)
            and invoke_evidence["evidence"].get("backend_accepted") is True,
            "semantic invoke omitted exact-target actor-owned evidence",
        )
        await wait_element(desktop, "Activation Count 1")
        marker(language, BEHAVIORS[3])

        click_result = await terminal(
            button.physical_click(
                lease,
                window=gtk_window,
                move_duration=0.25,
                postcondition={
                    "predicate": {"type": "exists"},
                    "timeout_ms": 3_000,
                    "allow_poll_fallback": True,
                },
            ),
            "smooth physical click",
        )
        click_outcome = outcome(click_result, "element_physical_click")
        click_evidence = click_outcome.get("result")
        require(
            isinstance(click_evidence, Mapping)
            and click_evidence.get("pointer_interpolated") is True,
            "physical click did not report interpolated pointer motion",
        )
        await wait_element(desktop, "Activation Count 2")
        marker(language, BEHAVIORS[4])

        await terminal(entry.set_text(""), "entry reset")
        text_result = await terminal(
            lease.keyboard.insert_text(
                UNICODE_TEXT,
                {
                    "target": "element",
                    "element": entry.ref,
                    "window_fallback": None,
                },
                strategy="auto",
                verify_length_only=False,
            ),
            "Unicode insertion",
        )
        text_evidence = outcome(text_result, "text_inserted").get("evidence")
        require(
            isinstance(text_evidence, Mapping)
            and text_evidence.get("selected_strategy") == "semantic"
            and text_evidence.get("utf8_bytes") == len(UNICODE_TEXT.encode("utf-8"))
            and text_evidence.get("unicode_scalars") == len(UNICODE_TEXT)
            and text_evidence.get("completed_scalars") == len(UNICODE_TEXT)
            and text_evidence.get("verified_length_only") is False,
            "Unicode insertion omitted exact delivery and strategy evidence",
        )
        marker(language, BEHAVIORS[5])

        failed = await button.invoke(
            postcondition={
                "predicate": {
                    "type": "state",
                    "state": "checked",
                    "value": True,
                },
                "timeout_ms": 750,
                "allow_poll_fallback": True,
            }
        )
        try:
            await failed.wait_until_terminal(10)
        except XenoteerError:
            pass
        else:
            raise RuntimeError("deliberately impossible postcondition unexpectedly succeeded")
        failed_result = failed.latest
        require(
            failed_result.get("lifecycle") == "failed"
            and failed_result.get("effect_stage")
            in {
                "semantic_action_dispatched",
                "semantic_state_changed",
            },
            "failed postcondition omitted visible-effect evidence",
        )
        await wait_element(desktop, "Activation Count 3")
        screenshot = await desktop.capture.screenshot(
            target={
                "kind": "window_visible",
                "window": gtk_window.ref,
                "coordinate_space": "frame",
            },
            include_cursor=True,
            max_bytes=4 * 1_048_576,
        )
        delivery = screenshot.get("delivery")
        require(
            isinstance(delivery, Mapping) and delivery.get("delivery") == "artifact",
            "failure screenshot was not retained as a private artifact",
        )
        screenshot_artifact = ArtifactRef.from_wire(
            delivery.get("artifact"),
            desktop_id=desktop.id,
            desktop_generation=desktop.generation,
            purpose="screenshot",
        )
        screenshot_bytes = await desktop.artifacts.download_bytes(screenshot_artifact)
        require(
            screenshot_bytes.startswith(b"\x89PNG\r\n\x1a\n")
            and hashlib.sha256(screenshot_bytes).hexdigest() == screenshot_artifact.sha256
            and screenshot.get("sha256") == screenshot_artifact.sha256,
            "failure screenshot bytes did not match their artifact evidence",
        )
        await desktop.artifacts.delete(screenshot_artifact)
        screenshot_artifact = None
        marker(language, BEHAVIORS[6])

        probe = desktop.submit({"type": "desktop_probe"})
        known_command_id = probe.id
        probe_result = await terminal(probe, "known-ID probe")
        require(outcome(probe_result, "probe").get("ready") is True, "probe was not ready")
        reconnect = await XenoteerClient.connect(
            ClientOptions(
                base_url=required("XENOTEER_API_BASE"),
                token=required("XENOTEER_TOKEN"),
                request_timeout=5,
            )
        )
        async with reconnect:
            recovered = await reconnect.desktop().command(known_command_id)
            recovered_result = await recovered.wait_until_terminal(10)
            require(
                recovered_result.get("command_id") == known_command_id
                and recovered_result.get("lifecycle") == "succeeded",
                "reconnect did not recover the known command ID",
            )
        marker(language, BEHAVIORS[7])

        old_window = xmessage_window
        old_identity = old_window.identity
        await terminate_process(desktop, process)
        process = None
        await desktop.windows.wait(
            {
                "target": {"type": "reference", "window": old_identity},
                "predicate": {"type": "gone"},
                "after_revision": None,
                "timeout_ms": 15_000,
            }
        )
        try:
            await old_window.snapshot()
        except XenoteerError as error:
            require(
                error.code == "stale_reference" or error.problem_code == "stale_reference",
                "old window failed with a non-stale error",
            )
        else:
            raise RuntimeError("old window reference remained valid after restart")
        process = await launch_xmessage(desktop, message)
        new_window = await wait_window(desktop, XMESSAGE_TITLE)
        require(new_window.identity != old_identity, "restart reused the exact window birth")
        marker(language, BEHAVIORS[8])

        ticket = await desktop.viewer.ticket(VIEWER_ORIGIN)
        metadata = ticket.metadata
        secret = ticket.expose_ticket()
        require(
            metadata.get("origin") == VIEWER_ORIGIN
            and metadata.get("mode") == "view_only"
            and metadata.get("audience") == "viewer_websocket"
            and metadata.get("use_policy") == "single_use"
            and 43 <= len(secret) <= 128
            and secret not in repr(ticket),
            "browser ticket was not exact-origin, single-use, and view-only",
        )
        marker(language, BEHAVIORS[9])
    except BaseException as error:
        operation_failure = error
    finally:
        failures = [] if operation_failure is None else [operation_failure]
        if screenshot_artifact is not None:
            try:
                await desktop.artifacts.delete(screenshot_artifact)
            except BaseException as error:
                failures.append(error)
        if process is not None:
            try:
                await terminate_process(desktop, process)
            except BaseException as error:
                failures.append(error)
        if len(failures) == 1:
            raise failures[0]
        if failures:
            raise BaseExceptionGroup(
                "behavior execution and resource cleanup failed",
                failures,
            )


async def exercise() -> None:
    verify_installed_origin()
    language = required("XENOTEER_QUICKSTART_LANGUAGE")
    require(
        re.fullmatch(r"[a-z-]+", language) is not None,
        "quick-start language label is invalid",
    )
    expect_authentication_failure = required("XENOTEER_EXPECT_AUTH_FAILURE") == "1"
    options = ClientOptions(
        base_url=required("XENOTEER_API_BASE"),
        token=required("XENOTEER_TOKEN"),
        request_timeout=5,
    )
    try:
        client = await XenoteerClient.connect(options)
    except XenoteerError as error:
        if expect_authentication_failure and error.code == "authentication" and error.status == 401:
            print(f"quickstart-ok language={language} mode=auth-failure")
            return
        raise
    async with client:
        require(
            not expect_authentication_failure,
            "invalid bearer unexpectedly authenticated",
        )
        require(
            client.negotiated_protocol.major == 1
            and client.negotiated_protocol.minor == 0
            and client.status.desktop.state == "ready",
            "server did not expose a ready frozen v1.0 desktop",
        )
        await exercise_connected(client, language)
        print(f"quickstart-ok language={language} mode=success")


def main() -> int:
    try:
        asyncio.run(asyncio.wait_for(exercise(), timeout=110))
    except XenoteerError as error:
        print(
            f"public Python behavior example failed: XenoteerError[{error.code}]",
            file=sys.stderr,
        )
        return 1
    except BaseExceptionGroup:
        print(
            "public Python behavior example failed: multiple bounded failures",
            file=sys.stderr,
        )
        return 1
    except (OSError, RuntimeError, asyncio.TimeoutError) as error:
        print(f"public Python behavior example failed: {error}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: packages/typescript/examples/phase6-behaviors.mjs
#!/usr/bin/env node
// SPDX-License-Identifier: Apache-2.0
//
// Executable installed-package qualification for the ten Phase 6 SDK
// behaviors. The artifact gate launches the GTK fixture from the exact derived
// image and supplies the four required XENOTEER_* connection variables.

import { createHash } from 'node:crypto';
import { isAbsolute, relative, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

import { XenoteerClient, XenoteerError } from '../dist/src/index.js';

const BEHAVIORS = Object.freeze([
  'status-capabilities',
  'scoped-lease-fixture-launch',
  'exact-window-element',
  'semantic-invoke',
  'smooth-physical-click-postcondition',
  'unicode-text-strategy',
  'screenshot-on-failure',
  'reconnect-known-command',
  'stale-reference-restart',
  'view-only-browser-ticket',
]);
const GTK_TITLE = 'Xenoteer GTK3 Fixture — Main';
const XMESSAGE_TITLE = 'xmessage';
const VIEWER_ORIGIN = 'https://viewer.example';
const UNICODE_TEXT = 'Xenoteer — العربية — 中文 — e\u0301 — 😀';
const TRANSPORT_REQUEST_TIMEOUT_MILLISECONDS = 35_000;
const SERVER_LONG_POLL_TIMEOUT_MILLISECONDS = 30_000;
// JavaScript promises do not provide structured cancellation. A Promise.race
// would report a timeout while an in-flight mutation or cleanup kept running.
// The package gate therefore owns the honest whole-process deadline.
const EXTERNAL_PROCESS_TIMEOUT_MILLISECONDS = 120_000;

function required(name) {
  const value = process.env[name];
  if (typeof value !== 'string' || value.length === 0)
    throw new Error(`required environment is missing: ${name}`);
  return value;
}

function requireCondition(condition, message) {
  if (!condition) throw new Error(message);
}

function verifyInstalledOrigin() {
  const expected = resolve(required('XENOTEER_EXPECTED_INSTALL_ROOT'));
  const installed = resolve(fileURLToPath(new URL('../dist/src/index.js', import.meta.url)));
  const within = relative(expected, installed);
  requireCondition(
    within === '' || (!within.startsWith('..') && !isAbsolute(within)),
    'TypeScript SDK resolved outside the staged npm installation',
  );
}

function marker(language, behavior) {
  console.log(`quickstart-ok language=${language} behavior=${behavior}`);
}

function windowSelector(title) {
  return {
    type: 'predicate',
    predicate: {
      type: 'text',
      field: 'title',
      matcher: { type: 'exact', value: title, case_sensitive: true },
    },
  };
}

function elementSelector(name) {
  return {
    scope: { type: 'desktop' },
    predicates: [
      {
        type: 'name',
        matcher: { type: 'exact', value: name, case_sensitive: true },
      },
    ],
    order: 'object_path_ascending',
    result_index: null,
  };
}

function accessibilitySpec(component = false) {
  return {
    expansion: {
      actions: false,
      value: false,
      text_metadata: false,
      text_content: false,
      attributes: false,
      relations: false,
      component,
    },
    limits: {
      max_visited_nodes: 25_000,
      max_depth: 64,
      max_matches: 10,
      timeout_ms: 10_000,
    },
  };
}

async function waitElement(desktop, name) {
  const selector = elementSelector(name);
  await desktop.accessibility.wait({
    target: { type: 'selector', selector, quantifier: 'exactly_one' },
    predicate: { type: 'exists' },
    after_revision: null,
    timeout_ms: SERVER_LONG_POLL_TIMEOUT_MILLISECONDS,
    allow_poll_fallback: true,
    ...accessibilitySpec(false),
  });
  return await desktop.accessibility.one(selector, accessibilitySpec(true));
}

async function waitWindow(desktop, title) {
  const selector = windowSelector(title);
  await desktop.windows.wait({
    target: { type: 'selector', selector, quantifier: 'exactly_one' },
    predicate: { type: 'exists' },
    after_revision: null,
    timeout_ms: SERVER_LONG_POLL_TIMEOUT_MILLISECONDS,
  });
  return await desktop.windows.one(selector, 'creation_ascending');
}

async function terminal(handle, label) {
  const result = await handle.waitUntilTerminal(20_000);
  requireCondition(result.lifecycle === 'succeeded', `${label} did not succeed`);
  return result;
}

function outcome(result, kind) {
  const value = result.outcome;
  requireCondition(
    value !== null && typeof value === 'object' && value.type === kind,
    `command omitted its ${kind} outcome`,
  );
  return value;
}

async function launchXmessage(desktop, message) {
  const result = await terminal(
    await desktop.applications.launch(
      'xmessage',
      [message],
    ),
    'xmessage launch',
  );
  const processRef = outcome(result, 'application_launched').process;
  requireCondition(
    processRef !== null && typeof processRef === 'object',
    'launch outcome omitted its exact process',
  );
  return processRef;
}

async function terminateProcess(desktop, processRef) {
  const result = await terminal(
    await desktop.applications.terminate(processRef, 2_000),
    'fixture termination',
  );
  const terminated = outcome(result, 'process_terminated').process;
  requireCondition(
    terminated !== null && typeof terminated === 'object' && terminated.state === 'exited',
    'fixture termination did not reap the exact process',
  );
}

async function exerciseConnected(client, language) {
  const desktop = client.desktop();
  const capabilityEntries = client.status.capabilities?.capabilities;
  requireCondition(Array.isArray(capabilityEntries), 'status omitted capabilities');
  const available = new Set(
    capabilityEntries
      .filter((entry) => entry?.status === 'available')
      .map((entry) => entry.id),
  );
  for (const capability of [
    'accessibility.atspi',
    'capture.screenshot',
    'input.pointer.smooth',
    'process.managed.terminate',
    'viewer.novnc.view_only',
    'window.observe.wait',
  ]) {
    requireCondition(
      available.has(capability),
      `fixture capability is unavailable: ${capability}`,
    );
  }
  marker(language, BEHAVIORS[0]);

  await desktop.withControl(60_000, async (lease) => {
    await exerciseScoped(desktop, language, lease);
  });
}

async function exerciseScoped(desktop, language, lease) {
  const message = `Xenoteer SDK Phase 6 — ${language}`;
  let processRef;
  let screenshotArtifact;
  let operationFailed = false;
  let operationFailure;
  try {
    processRef = await launchXmessage(desktop, message);
    marker(language, BEHAVIORS[1]);

    const xmessageWindow = await waitWindow(desktop, XMESSAGE_TITLE);
    const gtkWindow = await waitWindow(desktop, GTK_TITLE);
    const button = await waitElement(desktop, 'Stable Button');
    const entry = await waitElement(desktop, 'Stable Entry');
    requireCondition(
      JSON.stringify(xmessageWindow.identity) !== JSON.stringify(gtkWindow.identity),
      'exact windows aliased',
    );
    requireCondition(
      JSON.stringify(button.identity) !== JSON.stringify(entry.identity),
      'exact elements aliased',
    );
    const correlated = await xmessageWindow.snapshot();
    requireCondition(
      JSON.stringify(correlated.window.snapshot.process.managed_process)
        === JSON.stringify(processRef),
      'exact xmessage window did not correlate to the launched process',
    );
    marker(language, BEHAVIORS[2]);

    const invokeResult = await terminal(
      await button.invoke({ type: 'default' }),
      'semantic invoke',
    );
    const invokeEvidence = outcome(invokeResult, 'element_action').result;
    requireCondition(
      invokeEvidence?.operation === 'invoke'
        && JSON.stringify(invokeEvidence.element) === JSON.stringify(button.identity)
        && invokeEvidence.evidence?.backend_accepted === true,
      'semantic invoke omitted exact-target actor-owned evidence',
    );
    await waitElement(desktop, 'Activation Count 1');
    marker(language, BEHAVIORS[3]);

    const clickResult = await terminal(
      await button.click(gtkWindow.identity, {
        leaseId: lease.id,
        moveDurationMs: 250,
        postcondition: {
          predicate: { type: 'exists' },
          timeout_ms: 3_000,
          allow_poll_fallback: true,
        },
      }),
      'smooth physical click',
    );
    requireCondition(
      outcome(clickResult, 'element_physical_click').result?.pointer_interpolated === true,
      'physical click did not report interpolated pointer motion',
    );
    await waitElement(desktop, 'Activation Count 2');
    marker(language, BEHAVIORS[4]);

    await terminal(await entry.setText(''), 'entry reset');
    const textResult = await terminal(
      await lease.keyboard.insertText(
        UNICODE_TEXT,
        {
          target: 'element',
          element: entry.identity,
          window_fallback: null,
        },
        {
          strategy: 'auto',
          autoPolicy: {
            allowed_strategies: ['semantic'],
            fallback: 'before_effect_only',
          },
          semanticOptions: {
            insertion_point: { kind: 'caret' },
            selection: 'collapse_after',
            verify_length_only: false,
            postcondition: null,
          },
          clipboardOptions: null,
        },
      ),
      'Unicode insertion',
    );
    const textEvidence = outcome(textResult, 'text_inserted').evidence;
    requireCondition(
      textEvidence?.selected_strategy === 'semantic'
        && textEvidence.utf8_bytes === new TextEncoder().encode(UNICODE_TEXT).byteLength
        && textEvidence.unicode_scalars === [...UNICODE_TEXT].length
        && textEvidence.completed_scalars === [...UNICODE_TEXT].length
        && textEvidence.verified_length_only === false,
      'Unicode insertion omitted exact delivery and strategy evidence',
    );
    marker(language, BEHAVIORS[5]);

    const failed = await button.invoke(
      { type: 'default' },
      {
        postcondition: {
          predicate: { type: 'state', state: 'checked', value: true },
          timeout_ms: 750,
          allow_poll_fallback: true,
        },
      },
    );
    const failedResult = await failed.waitUntilTerminal(10_000);
    requireCondition(
      failedResult.lifecycle === 'failed'
        && ['semantic_action_dispatched', 'semantic_state_changed'].includes(
          failedResult.effect_stage,
        ),
      'deliberately failed postcondition omitted visible-effect evidence',
    );
    await waitElement(desktop, 'Activation Count 3');
    const screenshot = await desktop.capture.screenshot({
      target: {
        kind: 'window_visible',
        window: gtkWindow.identity,
        coordinate_space: 'frame',
      },
      format: 'png',
      include_cursor: true,
      region: null,
      scale: null,
      max_bytes: 4 * 1_048_576,
    });
    screenshotArtifact = screenshot.artifact;
    requireCondition(
      screenshotArtifact !== undefined,
      'failure screenshot was not retained as a private artifact',
    );
    const screenshotBytes = await screenshotArtifact.download({
      maxBytes: 4 * 1_048_576,
    });
    const digest = createHash('sha256').update(screenshotBytes).digest('hex');
    requireCondition(
      Buffer.from(screenshotBytes.subarray(0, 8)).equals(
        Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]),
      )
        && digest === screenshotArtifact.ref.sha256
        && digest === screenshot.result.sha256,
      'failure screenshot bytes did not match their artifact evidence',
    );
    await screenshotArtifact.delete();
    screenshotArtifact = undefined;
    marker(language, BEHAVIORS[6]);

    const probe = desktop.prepareSubmission({ type: 'desktop_probe' });
    const knownCommandId = probe.id;
    const probeResult = await terminal(await probe.send(), 'known-ID probe');
    requireCondition(
      outcome(probeResult, 'probe').ready === true,
      'probe did not report ready',
    );
    const reconnect = await XenoteerClient.connect({
      baseUrl: required('XENOTEER_API_BASE'),
      token: required('XENOTEER_TOKEN'),
      requestTimeoutMs: TRANSPORT_REQUEST_TIMEOUT_MILLISECONDS,
    });
    try {
      const recovered = await reconnect.desktop().command(knownCommandId);
      const recoveredResult = await recovered.waitUntilTerminal(10_000);
      requireCondition(
        recoveredResult.command_id === knownCommandId
          && recoveredResult.lifecycle === 'succeeded',
        'reconnect did not recover the known command ID',
      );
    } finally {
      await reconnect.close();
    }
    marker(language, BEHAVIORS[7]);

    const oldIdentity = xmessageWindow.identity;
    await terminateProcess(desktop, processRef);
    processRef = undefined;
    await desktop.windows.wait({
      target: { type: 'reference', window: oldIdentity },
      predicate: { type: 'gone' },
      after_revision: null,
      timeout_ms: 15_000,
    });
    try {
      await xmessageWindow.snapshot();
    } catch (error) {
      requireCondition(
        error instanceof XenoteerError
          && (error.code === 'stale_reference' || error.problemCode === 'stale_reference'),
        'old window failed with a non-stale error',
      );
    }
    requireCondition(xmessageWindow.stale, 'old window reference remained current after restart');
    processRef = await launchXmessage(desktop, message);
    const newWindow = await waitWindow(desktop, XMESSAGE_TITLE);
    requireCondition(
      JSON.stringify(newWindow.identity) !== JSON.stringify(oldIdentity),
      'restart reused the exact window birth',
    );
    marker(language, BEHAVIORS[8]);

    const ticket = await desktop.viewer.issueTicket(VIEWER_ORIGIN);
    const metadata = ticket.toJSON();
    const secret = ticket.consumeSecret();
    requireCondition(
      metadata.origin === VIEWER_ORIGIN
        && metadata.mode === 'view_only'
        && metadata.audience === 'viewer_websocket'
        && metadata.usePolicy === 'single_use'
        && typeof secret === 'string'
        && secret.length >= 32
        && !ticket.toString().includes(secret),
      'browser ticket was not exact-origin, single-use, and view-only',
    );
    marker(language, BEHAVIORS[9]);
  } catch (error) {
    operationFailed = true;
    operationFailure = error;
  } finally {
    const failures = operationFailed ? [operationFailure] : [];
    if (screenshotArtifact !== undefined) {
      try {
        await screenshotArtifact.delete();
      } catch (error) {
        failures.push(error);
      }
    }
    if (processRef !== undefined) {
      try {
        await terminateProcess(desktop, processRef);
      } catch (error) {
        failures.push(error);
      }
    }
    if (failures.length === 1)
      throw failures[0];
    if (failures.length > 1)
      throw new AggregateError(failures, 'behavior execution and resource cleanup failed');
  }
}

async function exercise() {
  verifyInstalledOrigin();
  requireCondition(
    EXTERNAL_PROCESS_TIMEOUT_MILLISECONDS >= 2 * TRANSPORT_REQUEST_TIMEOUT_MILLISECONDS,
    'external process deadline does not cover an operation plus cleanup',
  );
  const language = required('XENOTEER_QUICKSTART_LANGUAGE');
  requireCondition(/^[a-z-]+$/u.test(language), 'quick-start language label is invalid');
  const expectAuthenticationFailure = required('XENOTEER_EXPECT_AUTH_FAILURE') === '1';
  let client;
  try {
    client = await XenoteerClient.connect({
      baseUrl: required('XENOTEER_API_BASE'),
      token: required('XENOTEER_TOKEN'),
      requestTimeoutMs: TRANSPORT_REQUEST_TIMEOUT_MILLISECONDS,
    });
  } catch (error) {
    if (
      expectAuthenticationFailure
      && error instanceof XenoteerError
      && error.code === 'authentication'
      && error.status === 401
    ) {
      console.log(`quickstart-ok language=${language} mode=auth-failure`);
      return;
    }
    throw error;
  }
  try {
    requireCondition(
      !expectAuthenticationFailure,
      'invalid bearer unexpectedly authenticated',
    );
    requireCondition(
      client.negotiatedProtocol.major === 1
        && client.negotiatedProtocol.minor === 0
        && client.status.desktop.state === 'ready',
      'server did not expose a ready frozen v1.0 desktop',
    );
    await exerciseConnected(client, language);
    console.log(`quickstart-ok language=${language} mode=success`);
  } finally {
    await client.close();
  }
}

try {
  await exercise();
} catch (error) {
  const detail = error instanceof XenoteerError
    ? `${error.name}[${error.code}]`
    : error instanceof Error
      ? error.message
      : 'unknown safe failure';
  console.error(`public TypeScript behavior example failed: ${detail}`);
  process.exitCode = 1;
}
